src/bun.js/node/types.zig

```diff
@@ -1840,548 +1915,2971 @@ pub const Emitter = struct {
+            const deviceLen = _device.len;
+            bufOffset = if (_isAbsolute) deviceLen + 1 else deviceLen;
+            bufSize = bufOffset + tailLen;
+            if (tailLen > 0) {
+                // Move all bytes to the right by device.len + (maybe 1 for the separator).
+                // Use bun.copy because tail and buf overlap.
+                bun.copy(T, buf[bufOffset ..bufSize], tail);
+            }
+            @memcpy(buf[0 ..deviceLen], _device);
+            if (_isAbsolute) {
+                buf[deviceLen] = CHAR_BACKWARD_SLASH;
+            }
+        } else {
+            // Translated from the following JS code:
+            //   return isAbsolute ? `\\${tail}` : tail;
+            if (_isAbsolute) {
+                bufOffset = 1;
+                bufSize = bufOffset + tailLen;
+                if (tailLen > 0) {
+                    // Move all bytes to the right by 1 for the separator.
+                    // Use bun.copy because tail and buf overlap.
+                    bun.copy(T, buf[bufOffset ..bufSize], tail);
+                }
+                // Prepend the separator.
+                buf[0] = CHAR_BACKWARD_SLASH;
+            }
+        }
+
+        return buf[0 ..bufSize];
+    }
+
+    pub inline fn normalizePosix(path: []const u8, buf: []u8) []const u8 {
+        return normalizePosixT(u8, path, buf);
+    }
+
+    pub inline fn normalizeWindows(path: []const u8, buf: []u8) []const u8 {
+        return normalizeWindowsT(u8, path, buf);
+    }
+
+    pub fn normalize(globalObject: *JSC.JSGlobalObject, isWindows: bool, args_ptr: [*]JSC.JSValue, args_len: u16) callconv(.C) JSC.JSValue {
+        if (comptime is_bindgen) return JSC.JSValue.jsUndefined();
+        const path_ptr = if (args_len > 0) args_ptr[0] else JSC.JSValue.jsUndefined();
+         // Supress exeption in zig. It does globalThis.vm().throwError() in JS land.
+        validateString(globalObject, path_ptr, "path", .{}) catch {
+            return JSC.JSValue.jsUndefined();
+        };
+        const pathZStr = path_ptr.getZigString(globalObject);
+        const len = pathZStr.len;
+        if (len == 0) return toUTF8JSString(globalObject, CHAR_STR_DOT);
+
+        var stack_fallback = std.heap.stackFallback(
+            stack_fallback_size_small, 
+            JSC.getAllocator(globalObject)
+        );
+        const allocator = stack_fallback.get();
+
+        const pathZSlice = pathZStr.toSlice(allocator);
+        defer pathZSlice.deinit();
+
+        const bufLen = @max(len, MAX_PATH_BYTES);
+        if (bufLen > MAX_PATH_BYTES) {
+            const buf = allocator.alloc(u8, bufLen) catch bun.outOfMemory();
+            defer allocator.free(buf);
+            return if (isWindows) 
+                toJSString(globalObject, normalizeWindows(pathZSlice.slice(), buf))
+            else
+                toJSString(globalObject, normalizePosix(pathZSlice.slice(), buf));
+        }
+        var buf: PathBuffer = undefined;
```

Comment:
☝️ We only use the `allocator` when needed else we use the cheap stack `var buf: PathBuffer = undefined;`